ksqldb-api-client/src/main/java/io/confluent/ksql/api/client/impl/ClientImpl.java

```diff
@@ -133,22 +139,46 @@ public void close() {
       final Map<String, Object> properties,
       final ResponseHandlerSupplier<T> responseHandlerSupplier
   ) {
-
     final JsonObject requestBody = new JsonObject().put("sql", sql).put("properties", properties);
 
     final CompletableFuture<T> cf = new CompletableFuture<>();
+    makeRequest(
+        "/query-stream",
+        requestBody,
+        cf,
+        response -> handleSuccessfulQueryResponse(response, cf, responseHandlerSupplier)
+    );
+
+    return cf;
+  }
+
+  private CompletableFuture<Void> makeCloseQueryRequest(final String queryId) {
+    final CompletableFuture<Void> cf = new CompletableFuture<>();
+
+    makeRequest(
+        "/close-query",
+        new JsonObject().put("queryId", queryId),
+        cf,
+        response -> handleSuccessfulCloseQueryResponse(cf)
+    );
+
+    return cf;
+  }
 
+  private <T> void makeRequest(
+      final String path,
+      final JsonObject requestBody,
+      final CompletableFuture<T> cf,
+      final Handler<HttpClientResponse> successfulResponseHandler) {
     HttpClientRequest request = httpClient.request(HttpMethod.POST,
         serverSocketAddress, clientOptions.getPort(), clientOptions.getHost(),
-        "/query-stream",
-        response -> handleResponse(response, cf, responseHandlerSupplier))
+        path,
+        response -> handleResponse(response, cf, successfulResponseHandler))
```

Comment:
I don't think you need to pass in successfulResponseHandler here, you can just call cf.thenAccept(this::successfulResponseHandler) (or similar) in this method.

Reply:
Not quite. The CompletableFuture will succeed even if the response status is not 200. The successful response handler is only meant to be called if the response status is 200, else the error handling logic should be called. I realize the way this is currently set up is a bit convoluted, though, so I've refactored it to be more readable.